src/algorithms/reco/ParticleFlow.h

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Derek Anderson
+
+#ifndef EICRECON_PARTICLEFLOW_H
+#define EICRECON_PARTICLEFLOW_H
+
+#include <map>
+#include <string>
+#include <utility>
+#include <algorithm>
+#include <spdlog/spdlog.h>
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+// event data model definitions
+#include <edm4eic/Cluster.h>
+#include <edm4eic/TrackPoint.h>
+#include <edm4eic/TrackCollection.h>
+#include <edm4eic/TrackSegmentCollection.h>
+#include <edm4eic/ClusterCollection.h>
+#include <edm4eic/ReconstructedParticleCollection.h>
+// for algorithm configuration
+#include "algorithms/interfaces/WithPodConfig.h"
+#include "ParticleFlowConfig.h"
+
+namespace eicrecon{
+
+  // --------------------------------------------------------------------------
+  //! Particle Flow
+  // --------------------------------------------------------------------------
+  /*! This algorithm takes a collection of tracks (and their projections)
+   *  and a pair of calorimeter collections, from an ECal and an HCal,
+   *  and returns a collection of "particle flow objects", reconstructed
+   *  particles constructed from either tracks or calorimeter clusters.
+   *
+   *  As there will likely need to be multiple PF algorithms at any given
+   *  time, e.g. different eta regions may need to use different algorithms,
+   *  the particular choice of algorithm is set by a user-configurable
+   *  flag. The `process()` call will then run whichever algorithm is
+   *  specified.
+   */
+  class ParticleFlow : public WithPodConfig<ParticleFlowConfig> {
+
+    public:
+
+      // ----------------------------------------------------------------------
+      //! Projection Bundle
+      // ----------------------------------------------------------------------
+      /*! An intermediate struct to make calculations easier. Holds track
+       *  projections grouped together and the running sum of energy.
+       */
+      struct ProjectionBundle {
+        float energy = 0.;
+        std::vector<edm4eic::TrackPoint> projections;
+      };
+
+      // ----------------------------------------------------------------------
+      //! Merged Cluster
+      // ----------------------------------------------------------------------
+      /*! An intermediate struct to make calculations easier. Holds info on
+       *  calo clusters merged so far.
+       */
+      struct MergedCluster {
+        bool done = false;
+        int32_t pdg = 0;
+        float mass = 0.;
+        float charge = 0.;
+        float energy = 0.;
+        edm4hep::Vector3f momentum = {0., 0., 0.};
+        edm4hep::Vector3f weighted_position = {0., 0., 0.};
+        std::vector<edm4eic::Cluster> clusters;
```

Comment:
I'm worried about the copying that's going to be happening here. Do we really need a separate struct for this that we then need to copy into and out of? If we use podio types we already have, then we don't need to worry about that. Is there something missing in the podio cluster types?